File: macOS/FAME Smart Flasher/Services/Protocol/SLIPCodec.swift
import Foundation

/// SLIP (Serial Line Internet Protocol) encoder/decoder
/// Used for framing ESP32 bootloader packets
enum SLIPCodec {
    // SLIP special characters
    static let frameEnd: UInt8 = 0xC0
    static let frameEscape: UInt8 = 0xDB
    static let escapedEnd: UInt8 = 0xDC      // 0xC0 -> 0xDB 0xDC
    static let escapedEscape: UInt8 = 0xDD   // 0xDB -> 0xDB 0xDD

    /// Encode data with SLIP framing
    /// - Parameter data: Raw data to encode
    /// - Returns: SLIP-encoded packet with 0xC0 delimiters
    static func encode(_ data: Data) -> Data {
        var encoded = Data()
        encoded.reserveCapacity(data.count * 2 + 2)

        encoded.append(frameEnd)

        for byte in data {
            switch byte {
            case frameEnd:
                encoded.append(frameEscape)
                encoded.append(escapedEnd)
            case frameEscape:
                encoded.append(frameEscape)
                encoded.append(escapedEscape)
            default:
                encoded.append(byte)
            }
        }

        encoded.append(frameEnd)
        return encoded
    }

    /// Decode a SLIP-framed packet
    /// - Parameter slipPacket: SLIP-encoded packet (including delimiters)
    /// - Returns: Decoded raw data, or nil if invalid
    static func decode(_ slipPacket: Data) -> Data? {
        var decoded = Data()
        decoded.reserveCapacity(slipPacket.count)

        var inEscape = false
        var started = false

        for byte in slipPacket {
            if byte == frameEnd {
                if started && !decoded.isEmpty {
                    return decoded
                }
                started = true
                decoded.removeAll(keepingCapacity: true)
                continue
            }

            guard started else { continue }

            if inEscape {
                switch byte {
                case escapedEnd:
                    decoded.append(frameEnd)
                case escapedEscape:
                    decoded.append(frameEscape)
                default:
                    // Invalid escape sequence
                    decoded.append(byte)
                }
                inEscape = false
            } else if byte == frameEscape {
                inEscape = true
            } else {
                decoded.append(byte)
            }
        }

        return decoded.isEmpty ? nil : decoded
    }
}

/// Stateful SLIP decoder for streaming data
final class SLIPDecoder {
    private var buffer = Data()
    private var inEscape = false
    private var packetStarted = false

    /// Process incoming bytes and return complete packets
    /// - Parameter byte: Single byte to process
    /// - Returns: Complete decoded packet if one was received, nil otherwise
    func process(_ byte: UInt8) -> Data? {
        switch byte {
        case SLIPCodec.frameEnd:
            if packetStarted && !buffer.isEmpty {
                let packet = buffer
                reset()
                return packet
            }
            packetStarted = true
            buffer.removeAll(keepingCapacity: true)
            return nil

        case SLIPCodec.frameEscape:
            guard packetStarted else { return nil }
            inEscape = true
            return nil

        default:
            guard packetStarted else { return nil }

            if inEscape {
                inEscape = false
                switch byte {
                case SLIPCodec.escapedEnd:
                    buffer.append(SLIPCodec.frameEnd)
                case SLIPCodec.escapedEscape:
                    buffer.append(SLIPCodec.frameEscape)
                default:
                    buffer.append(byte)
                }
            } else {
                buffer.append(byte)
            }
            return nil
        }
    }

    /// Process multiple bytes at once
    /// - Parameter data: Data to process
    /// - Returns: Array of complete decoded packets
    func process(_ data: Data) -> [Data] {
        var packets: [Data] = []
        for byte in data {
            if let packet = process(byte) {
                packets.append(packet)
            }
        }
        return packets
    }

    /// Reset the decoder state
    func reset() {
        buffer.removeAll(keepingCapacity: true)
        inEscape = false
        packetStarted = false
    }
}

File: windows/Protocol/SLIPCodec.cs
// FAME Smart Flasher - C# Port
// Copyright 2025 Fyrby Additive Manufacturing & Engineering
// SPDX-License-Identifier: Proprietary

namespace FAMESmartFlasher.Protocol;

/// <summary>
/// SLIP (Serial Line Internet Protocol) encoder/decoder
/// Used for framing ESP32 bootloader packets
/// </summary>
public static class SLIPCodec
{
    // SLIP special characters
    private const byte FrameEnd = 0xC0;
    private const byte FrameEscape = 0xDB;
    private const byte EscapedEnd = 0xDC;      // 0xC0 -> 0xDB 0xDC
    private const byte EscapedEscape = 0xDD;   // 0xDB -> 0xDB 0xDD

    /// <summary>
    /// Encode data with SLIP framing
    /// </summary>
    /// <param name="data">Raw data to encode</param>
    /// <returns>SLIP-encoded packet with 0xC0 delimiters</returns>
    public static byte[] Encode(byte[] data)
    {
        var encoded = new List<byte>(data.Length * 2 + 2);

        encoded.Add(FrameEnd);

        foreach (var b in data)
        {
            switch (b)
            {
                case FrameEnd:
                    encoded.Add(FrameEscape);
                    encoded.Add(EscapedEnd);
                    break;
                case FrameEscape:
                    encoded.Add(FrameEscape);
                    encoded.Add(EscapedEscape);
                    break;
                default:
                    encoded.Add(b);
                    break;
            }
        }

        encoded.Add(FrameEnd);
        return encoded.ToArray();
    }

    /// <summary>
    /// Decode a SLIP-framed packet
    /// </summary>
    /// <param name="slipPacket">SLIP-encoded packet (including delimiters)</param>
    /// <returns>Decoded raw data, or null if invalid</returns>
    public static byte[]? Decode(byte[] slipPacket)
    {
        var decoded = new List<byte>(slipPacket.Length);
        var inEscape = false;
        var started = false;

        foreach (var b in slipPacket)
        {
            if (b == FrameEnd)
            {
                if (started && decoded.Count > 0)
                {
                    return decoded.ToArray();
                }
                started = true;
                decoded.Clear();
                continue;
            }

            if (!started) continue;

            if (inEscape)
            {
                switch (b)
                {
                    case EscapedEnd:
                        decoded.Add(FrameEnd);
                        break;
                    case EscapedEscape:
                        decoded.Add(FrameEscape);
                        break;
                    default:
                        // Invalid escape sequence
                        decoded.Add(b);
                        break;
                }
                inEscape = false;
            }
            else if (b == FrameEscape)
            {
                inEscape = true;
            }
            else
            {
                decoded.Add(b);
            }
        }

        return decoded.Count == 0 ? null : decoded.ToArray();
    }
}

/// <summary>
/// Stateful SLIP decoder for streaming data
/// </summary>
public class SLIPDecoder
{
    private readonly List<byte> _buffer = new();
    private bool _inEscape;
    private bool _packetStarted;

    /// <summary>
    /// Process incoming bytes and return complete packets
    /// </summary>
    /// <param name="byte">Single byte to process</param>
    /// <returns>Complete decoded packet if one was received, null otherwise</returns>
    public byte[]? Process(byte @byte)
    {
        switch (@byte)
        {
            case 0xC0: // FrameEnd
                if (_packetStarted && _buffer.Count > 0)
                {
                    var packet = _buffer.ToArray();
                    Reset();
                    return packet;
                }
                _packetStarted = true;
                _buffer.Clear();
                return null;

            case 0xDB: // FrameEscape
                if (!_packetStarted) return null;
                _inEscape = true;
                return null;

            default:
                if (!_packetStarted) return null;

                if (_inEscape)
                {
                    _inEscape = false;
                    switch (@byte)
                    {
                        case 0xDC: // EscapedEnd
                            _buffer.Add(0xC0);
                            break;
                        case 0xDD: // EscapedEscape
                            _buffer.Add(0xDB);
                            break;
                        default:
                            _buffer.Add(@byte);
                            break;
                    }
                }
                else
                {
                    _buffer.Add(@byte);
                }
                return null;
        }
    }

    /// <summary>
    /// Process multiple bytes at once
    /// </summary>
    /// <param name="data">Data to process</param>
    /// <returns>List of complete decoded packets</returns>
    public List<byte[]> Process(byte[] data)
    {
        var packets = new List<byte[]>();
        foreach (var b in data)
        {
            var packet = Process(b);
            if (packet != null)
            {
                packets.Add(packet);
            }
        }
        return packets;
    }

    /// <summary>
    /// Reset the decoder state
    /// </summary>
    public void Reset()
    {
        _buffer.Clear();
        _inEscape = false;
        _packetStarted = false;
    }
}
